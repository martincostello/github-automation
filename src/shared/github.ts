// Copyright (c) Martin Costello, 2023. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import fetch from 'node-fetch';
import { debug } from '@actions/core';
import { Context } from '@actions/github/lib/context';
// eslint-disable-next-line import/no-unresolved
import { Api } from '@octokit/plugin-rest-endpoint-methods/dist-types/types';
import { WorkflowConfig } from './config';

export async function getFileContents(octokit: Api, owner: string, repo: string, path: string, ref: string): Promise<string> {
  const { data: contents } = await octokit.rest.repos.getContent({
    owner,
    repo,
    path,
    ref,
  });
  const encoding = contents['encoding'];
  if (encoding === 'base64' && contents['content']) {
    return Buffer.from(contents['content'], 'base64').toString();
  } else if (encoding === 'none') {
    const response = await fetch(contents['download_url']);
    return await response.text();
  } else {
    throw new Error(`Unexpected encoding for ${path}: ${encoding}`);
  }
}

export async function getPull(
  octokit: Api,
  owner: string,
  repo: string,
  pull_number: number
): Promise<{
  html_url: string;
  mergeable_state: string;
  number: number;
}> {
  let pr = await octokit.rest.pulls.get({
    owner,
    repo,
    pull_number,
  });

  const logMergeableState = (): void => {
    debug(`${owner}/${repo}#${pull_number} mergeable_state: ${pr.data.mergeable_state}.`);
  };

  logMergeableState();

  // Poll for changes if the mergeable state is not yet known if a push just occurred.
  // The first read above will start a background job to re-calcuate the mergeability, but it may not be ready immediately.
  // See https://docs.github.com/en/rest/guides/using-the-rest-api-to-interact-with-your-git-database?apiVersion=2022-11-28#checking-mergeability-of-pull-requests
  // and https://github.com/pullreminders/backlog/issues/42#issuecomment-436412823.
  let pollCount = 0;
  const pollDelay = 5000;
  const timeout = 60000;
  const maxPollCount = timeout / pollDelay;

  while ((pr.data.mergeable_state === null || pr.data.mergeable_state === 'unknown') && pollCount < maxPollCount) {
    await new Promise((resolve) => setTimeout(resolve, pollDelay));
    pr = await octokit.rest.pulls.get({
      owner,
      repo,
      pull_number,
      // Specify cache headers to make the most of the GitHub API's rate limits.
      // See https://jamiemagee.co.uk/blog/making-the-most-of-github-rate-limits/.
      headers: {
        'If-Modified-Since': pr.headers['Last-Modified'],
        'If-None-Match': pr.headers['Etag'],
      },
    });

    logMergeableState();
    pollCount++;
  }

  return pr.data;
}

export type Repository = {
  full_name: string;
  repo: string;
  owner: string;
  default_branch: string;
  html_url: string;
};

type PaginatedApi = {
  octokit: Api & {
    paginate: import('@octokit/plugin-paginate-rest').PaginateInterface;
  };
};

export async function getReposForCurrentUser({ octokit }: PaginatedApi, type: 'owner' | 'member'): Promise<Repository[]> {
  const per_page = 100;
  const repos = await octokit.paginate(octokit.rest.repos.listForAuthenticatedUser, {
    per_page,
    type,
  });

  return repos
    .filter((repo) => !repo.archived)
    .filter((repo) => !repo.fork)
    .filter((repo) => !repo.is_template)
    .map((repo) => {
      return {
        full_name: repo.full_name,
        repo: repo.name,
        owner: repo.owner.login,
        default_branch: repo.default_branch,
        html_url: repo.html_url,
      };
    });
}

export async function getWorkflowConfig(octokit: Api, context: Context): Promise<WorkflowConfig> {
  return JSON.parse(await getFileContents(octokit, context.repo.owner, context.repo.repo, '.github/workflow-config.json', context.sha));
}

export type RepositoryDotNetSdk = {
  version: string;
  line: number;
};

type GlobalJson = {
  sdk: {
    version: string;
  };
};

export async function getDotNetSdk(octokit: Api, owner: string, repo: string, ref: string): Promise<RepositoryDotNetSdk | null> {
  let globalJsonString: string;

  try {
    globalJsonString = await getFileContents(octokit, owner, repo, 'global.json', ref);
  } catch (err) {
    return null;
  }

  const globalJson: GlobalJson = JSON.parse(globalJsonString);
  const sdkVersion = globalJson.sdk?.version;

  let lineNumber = -1;

  if (sdkVersion) {
    const globalJsonLines = globalJsonString.split('\n');
    for (let i = 0; i < globalJsonLines.length; i++) {
      const line = globalJsonLines[i];
      if (line.includes(sdkVersion)) {
        lineNumber = i + 1;
        break;
      }
    }
  }

  return {
    version: sdkVersion,
    line: lineNumber,
  };
}
