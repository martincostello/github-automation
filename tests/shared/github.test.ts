// Copyright (c) Martin Costello, 2023. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import { getOctokit as getClient } from '@actions/github';
import { beforeAll, beforeEach, describe, expect, jest, test } from '@jest/globals';
import {
  getDotNetSdk,
  getFileContents,
  getPull,
  getReposForCurrentUser,
  getUpdateConfiguration,
  getWorkflowConfig,
} from '../../src/shared/github';
import { setup } from '../fixtures';

const owner = 'owner';
const repo = 'repo';
const ref = 'main';

const getOctokit = (token: string = 'fake-token') => getClient(token);

describe('getFileContents', () => {
  const path = 'some/file';

  describe('when the content is base64 encoded', () => {
    let octokit;

    beforeEach(async () => {
      await setup('getFileContents/base64');
      octokit = getOctokit();
    });

    test('returns the decoded contents', async () => {
      const actual = await getFileContents(octokit, owner, repo, path, ref);
      expect(actual).toBe('Hello world!');
    });
  });

  describe('when the content is a URL', () => {
    let octokit;

    beforeEach(async () => {
      await setup('getFileContents/url');
      octokit = getOctokit();
    });

    test('returns the contents', async () => {
      const actual = await getFileContents(octokit, owner, repo, path, ref);
      expect(actual).toBe('Hello world!');
    });
  });

  describe('when the content has an unknown encoding', () => {
    let octokit;

    beforeEach(async () => {
      await setup('getFileContents/unknown');
      octokit = getOctokit();
    });

    test('throws an error', async () => {
      await expect(getFileContents(octokit, owner, repo, path, ref)).rejects.toThrow('Unexpected encoding for some/file: foo');
    });
  });
});

describe('getUpdateConfiguration', () => {
  describe('when the file exists', () => {
    let octokit;

    beforeEach(async () => {
      await setup('getUpdateConfiguration/exists');
      octokit = getOctokit();
    });

    test('returns the configuration', async () => {
      const actual = await getUpdateConfiguration(octokit, owner, repo, ref);
      expect(actual).not.toBeUndefined();
      expect(actual).not.toBeNull();
      expect(actual?.ignore).toBe(true);
      expect(actual?.['exclude-nuget-packages']).toBe('excluded');
      expect(actual?.['include-nuget-packages']).toBe('included');
      expect(actual?.['update-nuget-packages']).toBe(true);
    });
  });

  describe('when the file does not exist', () => {
    let octokit;

    beforeEach(async () => {
      await setup('getUpdateConfiguration/not-found');
      octokit = getOctokit();
    });

    test('returns null', async () => {
      const actual = await getUpdateConfiguration(octokit, owner, repo, ref);
      expect(actual).toBeNull();
    });
  });
});

describe('getWorkflowConfig', () => {
  let octokit;

  beforeEach(async () => {
    await setup('getWorkflowConfig/exists');
    octokit = getOctokit();
  });

  test('returns the configuration', async () => {
    const context = {
      repo: {
        owner,
        repo,
      },
      sha: ref,
    };
    const actual = await getWorkflowConfig(octokit, context as any);
    expect(actual).not.toBeUndefined();
    expect(actual).not.toBeNull();
    expect(actual?.checksOfInterest).toStrictEqual(['a', 'b']);
  });
});

describe('getDotNetSdk', () => {
  describe('when the file exists', () => {
    let actual;

    beforeAll(async () => {
      await setup('getDotNetSdk/exists');
      const octokit = getOctokit();
      actual = await getDotNetSdk(octokit, owner, repo, ref);
    });

    test('returns the version', () => {
      expect(actual?.version).toBe('7.0.400');
    });

    test('returns the line number', () => {
      expect(actual?.line).toBe(3);
    });
  });

  describe('when there is no SDK version', () => {
    let actual;

    beforeAll(async () => {
      await setup('getDotNetSdk/no-version');
      const octokit = getOctokit();
      actual = await getDotNetSdk(octokit, owner, repo, ref);
    });

    test('returns no version', () => {
      expect(actual?.version).toBeUndefined();
    });

    test('returns no line number', () => {
      expect(actual?.line).toBe(-1);
    });
  });

  describe('when the file does not exist', () => {
    let octokit;

    beforeEach(async () => {
      await setup('getDotNetSdk/not-found');
      octokit = getOctokit();
    });

    test('returns null', async () => {
      const actual = await getDotNetSdk(octokit, owner, repo, ref);
      expect(actual).toBeNull();
    });
  });
});

describe('getReposForCurrentUser', () => {
  describe.each([['all'], ['member'], ['owner']])('when there are repositories for %s', (type: string) => {
    let actual;

    beforeAll(async () => {
      await setup('getReposForCurrentUser/some');
      const octokit = getOctokit(`${type}-token`);
      actual = await getReposForCurrentUser(octokit, type as any);
    });

    test('returns the correct repositories', async () => {
      expect(actual).not.toBeNull();
      expect(actual.length).not.toBe(0);
      expect(actual).toMatchSnapshot();
    });
  });

  describe.each([['all'], ['member'], ['owner']])('when there are no repositories for %s', (type: string) => {
    let actual;

    beforeAll(async () => {
      await setup('getReposForCurrentUser/none');
      const octokit = getOctokit();
      actual = await getReposForCurrentUser(octokit, type as any);
    });

    test('returns an empty array', async () => {
      expect(actual).not.toBeNull();
      expect(actual.length).toBe(0);
    });
  });
});

describe('getPull', () => {
  const getOctokitForPulls = (responses: any | any[]): any => {
    let mock = jest.fn();

    if (!Array.isArray(responses)) {
      responses = [responses];
    }

    for (const data of responses) {
      mock = mock.mockReturnValueOnce({
        data,
        headers: {
          'Etag': '42',
          'Last-Modified': new Date().toUTCString(),
        },
      });
    }

    return {
      rest: {
        pulls: {
          get: mock,
        },
      },
    };
  };

  describe.each([
    [[false], false],
    [[true], true],
    [[null, false], false],
    [[null, true], true],
  ])('when mergeable is %s', (mergeables: any[], expected: boolean) => {
    let actual;

    beforeEach(async () => {
      const octokit = getOctokitForPulls(mergeables.map((mergeable) => ({ mergeable })));
      actual = await getPull(octokit, owner, repo, 42);
    }, 15000);

    test('returns mergeable', async () => {
      expect(actual).not.toBeNull();
      expect(actual['mergeable']).toBe(expected);
    });
  });
});
