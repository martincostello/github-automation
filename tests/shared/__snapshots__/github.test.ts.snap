// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`getReposForCurrentUser when there are repositories for all returns the correct repositories 1`] = `
[
  {
    "default_branch": "main",
    "full_name": "org/name",
    "html_url": "https://github.com/org/name",
    "owner": "org",
    "repo": "name",
  },
  {
    "default_branch": "develop",
    "full_name": "org2/name2",
    "html_url": "https://github.com/org2/name2",
    "owner": "org2",
    "repo": "name2",
  },
  {
    "default_branch": "main",
    "full_name": "org3/name2",
    "html_url": "https://github.com/org3/name2",
    "owner": "org3",
    "repo": "name2",
  },
  {
    "default_branch": "main",
    "full_name": "org3/name3",
    "html_url": "https://github.com/org3/name3",
    "owner": "org3",
    "repo": "name3",
  },
  {
    "default_branch": "main",
    "full_name": "org3/name4",
    "html_url": "https://github.com/org3/name4",
    "owner": "org3",
    "repo": "name4",
  },
]
`;

exports[`getReposForCurrentUser when there are repositories for member returns the correct repositories 1`] = `
[
  {
    "default_branch": "main",
    "full_name": "org/name",
    "html_url": "https://github.com/org/name",
    "owner": "org",
    "repo": "name",
  },
  {
    "default_branch": "develop",
    "full_name": "org2/name2",
    "html_url": "https://github.com/org2/name2",
    "owner": "org2",
    "repo": "name2",
  },
  {
    "default_branch": "main",
    "full_name": "org3/name2",
    "html_url": "https://github.com/org3/name2",
    "owner": "org3",
    "repo": "name2",
  },
  {
    "default_branch": "main",
    "full_name": "org3/name3",
    "html_url": "https://github.com/org3/name3",
    "owner": "org3",
    "repo": "name3",
  },
  {
    "default_branch": "main",
    "full_name": "org3/name4",
    "html_url": "https://github.com/org3/name4",
    "owner": "org3",
    "repo": "name4",
  },
]
`;

exports[`getReposForCurrentUser when there are repositories for owner returns the correct repositories 1`] = `
[
  {
    "default_branch": "main",
    "full_name": "org/name",
    "html_url": "https://github.com/org/name",
    "owner": "org",
    "repo": "name",
  },
  {
    "default_branch": "develop",
    "full_name": "org2/name2",
    "html_url": "https://github.com/org2/name2",
    "owner": "org2",
    "repo": "name2",
  },
  {
    "default_branch": "main",
    "full_name": "org3/name2",
    "html_url": "https://github.com/org3/name2",
    "owner": "org3",
    "repo": "name2",
  },
  {
    "default_branch": "main",
    "full_name": "org3/name3",
    "html_url": "https://github.com/org3/name3",
    "owner": "org3",
    "repo": "name3",
  },
  {
    "default_branch": "main",
    "full_name": "org3/name4",
    "html_url": "https://github.com/org3/name4",
    "owner": "org3",
    "repo": "name4",
  },
]
`;
